### PR TITLE
audio: avoid blocking in pull AOs

### PR DESCRIPTION
**Pull request by t-8ch**
Continuation of #12643 

Only tested on ao_pipewire.

@handlerug Could you test it with ao_coreaudio?
@tmm1 Could you test it with ao_audiotrack?

### CODE REVIEW

File: audio/out/internal.h [hunk 1]
```diff
@@ -201,6 +201,8 @@ struct ao_driver {
 // These functions can be called by AOs.
 
 int ao_read_data(struct ao *ao, void **data, int samples, int64_t out_time_ns);
+__attribute__((warn_unused_result))
```
Comment (N-R-K) [inline audio/out/internal.h:204]: I think nearly all the `__attribute` usage is currently guarded via ifdefs or similar. I'd add `MP_WARN_UNUSED` or something similar to `osdep/compiler.h`.

```diff
--- a/osdep/compiler.h
+++ b/osdep/compiler.h
@@ -7,10 +7,12 @@
 #define PRINTF_ATTRIBUTE(a1, a2) __attribute__ ((format(printf, a1, a2)))
 #define MP_NORETURN __attribute__((noreturn))
 #define MP_FALLTHROUGH __attribute__((fallthrough))
+#define MP_WARN_UNUSED __attribute__((warn_unused_result))
 #else
 #define PRINTF_ATTRIBUTE(a1, a2)
 #define MP_NORETURN
 #define MP_FALLTHROUGH do {} while (0)
+#define MP_WARN_UNUSED
 #endif
```

Comment (t-8ch) [reply audio/out/internal.h:204]: Not all of them:
```
ta/ta.h
24:#define TA_PRF(a1, a2) __attribute__ ((format(printf, a1, a2)))
34:#define TA_PRF(a1, a2) __attribute__ ((format (gnu_printf, a1, a2)))
libmpv/client.h
33:#define MPV_EXPORT __attribute__((visibility("default")))
```

And a few more. If it is preferred I can change it though.

Comment (N-R-K) [reply audio/out/internal.h:204]: Both the example is under ifdefs and has a fallback though (which was my main point, doing it in `compiler.h` isn't necessary but it makes more sense than inlining it).

https://github.com/mpv-player/mpv/blob/84de84b8aa6023efb3f8fac368d802359657268c/ta/ta.h#L23-L29

https://github.com/mpv-player/mpv/blob/84de84b8aa6023efb3f8fac368d802359657268c/libmpv/client.h#L29-L38

Comment (t-8ch) [reply audio/out/internal.h]: Indeed, update!

**Comment by N-R-K** [comment]
The CI is failing on f5c3b34383545f29a455570bda7ac060633e6e8a because it's using `pthread_*` directly. mpv uses the `mp_thread` wrappers since (*checks commit time of 3a8b107f6216b38a*) 5 hours ago.

Just renaming `pthread_mutex_*` to `mp_mutex_*` should do the job.

**Review by sfan5 — APPROVED** [review]
lgtm

**Review by philipl — APPROVED** [review]

File: audio/out/buffer.c [hunk 2]
```diff
@@ -181,21 +205,26 @@ static int read_buffer(struct ao *ao, void **data, int samples, bool *eof)
 int ao_read_data(struct ao *ao, void **data, int samples, int64_t out_time_ns)
 {
     struct buffer_state *p = ao->buffer_state;
-    assert(!ao->driver->write);
 
     mp_mutex_lock(&p->lock);
 
-    int pos = read_buffer(ao, data, samples, &(bool){0});
+    int pos = ao_read_data_unlocked(ao, data, samples, out_time_ns, true);
```
Comment (tmm1) [inline audio/out/buffer.c:211]: wouldn't this be better named `ao_read_data_locked`, since the lock is take in the previous line?

Comment (t-8ch) [reply audio/out/buffer.c:211]: This does indeed better match other similar uses in mpv.
My personal intuition is different, but I'll send a PR.

**Pull request merged**